source/vibe/core/sync.d

```diff
@@ -110,6 +110,104 @@ unittest {
 	});
 }
 
+/** Semaphore implementation for tasks.
+
+	It is not thread-safe, which is on purpose for performance reasons.
+	It will lock up to an adjustable maximum count of internal tasks.
+	Usage example is to limit concurrent connections in a connection pool
+*/
+class LocalTaskSemaphore
+{
+	// requires a queue
+	import std.container.binaryheap;
+	import std.container.array;
+	import vibe.utils.memory;
+
+	private {
+		struct Waiter {
+			ManualEvent signal;
+			ubyte priority;
+			uint seq;
+		}
+
+		BinaryHeap!(Array!Waiter, asc) m_waiters;
+		uint m_maxLocks;
+		uint m_locks;
+		uint m_seq;
+	}
+
+	@property void maxLocks(uint max_locks) { m_maxLocks = max_locks; }
+	@property uint maxLocks() const { return m_maxLocks; }
+	@property uint available() const { return m_maxLocks - m_locks; }
```

Comment:
This could use some annotations, at the very lest `nothrow`. `@safe` and `pure` are also candidate, but since we don't use them anywhere else...